### DISCUSSION TITLE
Programmaticaly trigger a screenshot

### DISCUSSION THREAD

**Discussion by Datoh** [17cupsofcoffee/tetra#251] [Q&A] [answered]
Hi,

How I get a screenshot programmaticaly.
My goal is to build an animated gif of the game.

Regards,
Datoh.

**Comment by 17cupsofcoffee** [comment]
This would require:

* Drawing the scene to a `Canvas` (currently supported)
* Transferring the data from the `Canvas`, which lives on the GPU, into an `ImageData`, which lives on the CPU (not currently supported)
* Reading the data contained in the `ImageData` (not currently supported)

I definitely want to make this possible in the future, I just need to make sure the API is right/safe.

**Reply by Datoh** [reply]
I implemented the first part and I saw that it is impossible to get data from texture/canvas but it is possible to set them

**Reply by Datoh** [reply]
Do you think I will be difficult for me to add this fetaure to Tetra?

**Reply by 17cupsofcoffee** [reply]
I don't think it'll be too difficult, I might try it tonight 🙂 

The main thing I want to fix before I implement this functionality is the fact that some areas of the code use `Texture`s that contain uninitialized memory (specifically canvases and fonts). I don't want to let the user read that garbage data back to the CPU, so I need to make sure all `Texture`s are initialized before the user gets access to them.

After that's fixed, I should be able to use `glGetTexImage` to pull the texture data down into an `ImageData`, and then I can expose some methods on `ImageData` for dumping the data/saving to a file.

If I don't get time to do this myself, I'll make an issue with some more detailed instructions so someone else can PR it :+1:

**Comment by 17cupsofcoffee** [comment, marked as answer]
The `main` branch now has support for turning a `Texture`/`Canvas` into an `ImageData` via a `get_data` method, and you can retrieve the generated RGBA data via `ImageData::into_rgba8`. API subject to change based on feedback/how I feel about it when I'm less tired 😄 

I may add some helper methods for saving screenshots in the future, but this should give you enough functionality to get started, at least.

**Reply by Datoh** [reply]
Thx a lot !

**Reply by 17cupsofcoffee** [reply]
No problem!
